File: backend/src/main/java/com/example/cellphonesclone/models/Product.java
package com.example.cellphonesclone.models;

import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(name = "products")
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Builder
public class Product extends BaseEntity{
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long productId;

    @Column(name = "name", length = 350, nullable = false)
    private String name;

    @Column(name = "price", nullable = false)
    private Float price;

    @Column(name = "thumbnail", length = 300)
    private String thumbnail;

    @Column(name = "description")
    private String description;

    @Column(name = "in_stock")
    private Integer inStock;

    @Column(name = "operating_system", length = 255)
    private String operatingSystem;

    @Column(name = "screen_size", precision = 5, scale = 2)
    private BigDecimal screenSize;

    @Column(name = "battery_capacity")
    private Integer batteryCapacity;

    @Column(name = "ram")
    private Integer ram;

    @Column(name = "rom")
    private Integer rom;

    @Column(name = "front_camera", length = 350)
    private String frontCamera;

    @Column(name = "main_camera", length = 350)
    private String mainCamera;

    @Column(name = "color", length = 255)
    private String color;

    @Column(name = "release_date")
    private Integer releaseDate;

    @ManyToOne
    @JoinColumn(name = "brands_id")
    private Brand brand;

    @JsonManagedReference
    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    private List<ProductImage> productImages;

    public ProductImage getThumbnail() {
        return productImages != null && !productImages.isEmpty()
                ? productImages.get(0)
                : null;
    }
}

File: frontend/cellphonesclone-angular/src/app/models/product.ts
import { ProductImage } from "./product.image";
export interface Product {
  productId: number;
  name: string;
  price: number;
  thumbnail: string;
  description: string;
  category_id: number;
  url: string;
  product_images: ProductImage[];
  operating_system?: string; // Optional, based on backend field
  screen_size?: number; // BigDecimal should be converted to number in JS/TS
  battery_capacity?: number;
  ram?: number;
  rom?: number;
  front_camera?: string;
  main_camera?: string;
  color?: string;
  release_date?: number; // Year
  in_stock?: number;
  brand_id?: number;
}

export function mapJsonToProduct(data: any): Product {
  return {
    productId: data.productId, // Map productId to id
    name: data.name,
    price: data.price,
    thumbnail: data.thumbnail,
    description: data.description,
    category_id: data.brand?.id, // Assuming you want to use brand_id as category_id
    url: data.thumbnail, // Or generate a URL if needed
    product_images: data.productImages.map((img: any) => ({
      id: img.id,
      imageUrl: img.imageUrl
    })),
    operating_system: data.operatingSystem,
    screen_size: data.screenSize,
    battery_capacity: data.batteryCapacity,
    ram: data.ram,
    rom: data.rom,
    front_camera: data.frontCamera,
    main_camera: data.mainCamera,
    color: data.color,
    release_date: data.releaseDate,
    in_stock: data.inStock,
    brand_id: data.brand?.id
  };
}